### ISSUE TITLE
Include a "Journey Test" in the CI build of each track

### ISSUE THREAD

**Issue by jtigger** [exercism/discussions#81]
_(spawned from discussion in #79)_

**tl;dr:** come up with a way to automate [journey style](http://martinfowler.com/bliki/UserJourneyTest.html) tests in CI for each track so that maintainers spend less time doing bookkeeping and have more room to use their judgment and creativity.
# Context

One of the more labor intensive parts of maintaining a track is ensuring that submissions from contributors are fit for production use.  The maintainer must verify:
- the tests and sample implementation are present and fully implement the exercise (especially with respect to the attendant `canonical-data.json` or, absent that on par with implementations in other tracks);
- the tests are reasonably exemplary: especially how those tests define the interface/protocols of the solution;
- if it's a new-to-the-track exercise, that it was properly included in the `config.json`;
- that the exercise can be successful served by the CLI (e.g. does not include any binaries);
- in the case where the build of the exercise _outside_ the track repo runs differently than the exercise _inside_ the track repo, that when fetched by exercism programmers, will continue to work.
# Problem

Some of this work requires human judgment, but a good deal of it is manual, monotonous and error-prone.  It also takes precious volunteer time of maintainers on ensuring that newly implemented exercises don't break production.
# Proposal

In CI (i.e. Travis) automate a full [Journey Test](http://martinfowler.com/bliki/UserJourneyTest.html) of an exercism programmer implementing the all the exercises of the track, for each track.  This involves:
1. Setup both the track language tool chain and Ruby in CI;
2. Download and stand-up `x-api`, configured to serve exercises from the HEAD of the branch being built instead of the ref of the submodule;
3. Download and configure the exercism CLI;
4. Iterate over the set of exercises, fetch each, "solve" by copying over the "example" solution and run the tests.
## Requirement: Be Modular

This should be done in a way such that the work required specifically for the track (e.g. which track is in play, where to copy solutions from, how to run the tests on the exercises, what pre- or post- steps are required) is separated from the common work of standing-up a local copy of `x-api`.  In this way, whenever details about the common work changes, it can be reflected in the Journey in one place.
# Resources
- The Java track has implemented such a test in bash: [journey-test.sh](https://github.com/exercism/xjava/blob/master/bin/journey-test.sh).
- Regardless of the track language, Ruby is required in the CI environment as it is a dependency of `x-api` (read: given the desire to provide a test harness and the difficulty in maintaining bash, implementers are strongly encouraged to consider Ruby).
# Next Steps
- Question the premise.  Question the proposal.  Is there another/easier way to squash the waste of reviewing the syntax of track submissions?
- If you agree with the proposal, use the prior art to guide an implementation in one track.  Be mindful of the parts that are track-specific.
- Use the test harness to implement the same on a second track and evaluate.
- Incorporating learnings, if this approach seems sound, fan-out an issue in each track for implementing this in Travis, providing implementation guidance.

**Comment by petertseng** [comment]
One thing that will help me to evaluate the benefits is to answer the question: What additional checks does this give us that are not already provided by running the tests in each track plus configlet?

> does not include any binaries

This seems to be one answer to the above question

> in the case where the build of the exercise *outside* the track repo runs differently than the exercise *inside* the track repo

I would like to know in what cases this happens - can it be solved by changing the track's existing test suite?

An additional point I can add to this list: It is currently possible to add a nonexistent exercise, and configlet is okay with this: https://github.com/exercism/xrust/pull/207

**Comment by ryanplusplus** [comment]
@petertseng I have the same concern regarding the benefits of doing this instead of just testing inside the track. It feels like an awfully complicated way to address issues that we don't have or that could be addressed in a simpler way. For example, can't we catch the non-existent exercise example given above with a fairly straightforward change to configlet?

**Comment by kytrinyx** [comment]
The thing that this changes which configlet cannot, is the integration with our fetch and submission APIs.

We've had things that pass the basic checks, and then blow up because of some unforeseen thing that the API can't handle.

I am working on an integration test that can run before I deploy the API, which will check every track against every API endpoint, but this check happens much later in the process. It would be nice to catch these things when the person submitting the pull request is in the middle of it, not when I'm deploying the site.

**Comment by kytrinyx** [comment]
> It is currently possible to add a nonexistent exercise, and configlet is okay with this [...] but I presume this mistake would be caught by the journey test.

Yeah, this would be caught by my tests in the API before it gets deployed.

**Comment by ryanplusplus** [comment]
@kytrinyx can you elaborate on the type of thing that could pass the basic tests but then blow up on the API? I'm not super familiar with the API side of things, but it feels pretty bounded from the track standpoint so I'm having a hard time understanding where things might go wrong.

**Comment by kytrinyx** [comment]
The typical thing is misspelling a problem name, e.g. `braket-push` instead of `bracket-push`. That will pass configlet so long as the slug in the config file matches the directory name of the exercise.

We could pass a flag to configlet so it checks valid problem names via the api, though.

Once we had a zipped archive file, and that totally blew up because the files are added as text to JSON and served as application/json from the endpoint.

**Comment by jtigger** [comment]
In terms of what such a test would cover is nominally the intersection of the config+content of a track and the code-path from `exercism fetch` through to an exercise being delivered from x-api.

**Comment by jtigger** [comment]
> > in the case where the build of the exercise outside the track repo runs differently than the exercise inside the track repo

> I would like to know in what cases this happens - can it be solved by changing the track's existing test suite?

Excellent question.

In the Java track (and by way of C&P, Kotlin), we've got a Gradle multi-module build: a root module (in the `exercises`) and each individual exercise is a sub-module.

Gradle scans up directories for parent build config.  When building an exercise within the `xjava` repo, this build config is modified: a) the source set is reset from `src/main/java` to `src/example/java`; and b) all of the tests an enabled (by default only the first is enabled).

When the exercise is delivered (i.e. no longer has that parent build config), the exercise builds using the expected source set and with all but the first test marked as "skipped."

**Comment by jtigger** [comment]
@petertseng said, above:

> An additional point I can add to this list: It is currently possible to add a nonexistent exercise, and configlet is okay with this: exercism/xrust#207 , but I presume this mistake would be caught by the journey test.

Yes.

**Comment by jtigger** [comment]
@kytrinyx said, above:

>I am working on an integration test that can run before I deploy the API, which will check every track against every API endpoint, but this check happens much later in the process. It would be nice to catch these things when the person submitting the pull request is in the middle of it, not when I'm deploying the site.

How often does something like this come up?  Have you had to not deploy or revert a track because of something like this?

**Comment by kytrinyx** [comment]
> How often does something like this come up?

@jtigger It's not very often, every few weeks, maybe?

> Have you had to not deploy or revert a track because of something like this?

Yeah, both of the above. I'm working on better tests that I can run before a deploy.

**Comment by jtigger** [comment]
I will say, having lived on the other side of the implementation of this feature, it's brought peace-of-mind that whatever changes happen in the CLI and x-api, any new submissions to this track really only need convention, style and quality review.  This kind of check not only means it takes less time for me to review submissions, but by extension, easier to be responsive.

**Comment by kytrinyx** [comment]
> having lived on the other side of the implementation of this feature, it's brought peace-of-mind that whatever changes happen in the CLI and x-api, any new submissions to this track really only need convention, style and quality review

This sounds wonderful, to be honest. Having been behind many of the CLI and x-api changes...

**Comment by jtigger** [comment]
FYI: just added this to the Kotlin track exercism/xkotlin#33.

I'd like to emphasize how, in the sum, I've regained uncounted _hours_ that I've channelled into more thoughtful discussions over API design rather than mind-numbing double-checking plumbing... problems I'd rather be solving. :)

**Comment by kytrinyx** [comment]
_We've imported this issue to the https://github.com/exercism/exercism.io repository._
